src/processors/BaseProcessor.h

```diff
@@ -40,16 +48,72 @@ struct ConnectionInfo
+using PortTypesVector = chowdsp::SmallVector<PortType, 4>;
+
+template <typename Port, typename PortMapper>
+static PortTypesVector initialisePortTypes (PortMapper mapper)
+{
+    auto portTypes = PortTypesVector (magic_enum::enum_count<Port>(), PortType::audio);
+    magic_enum::enum_for_each<Port> ([&portTypes, &mapper] (auto portType)
+                                     {
+                                         const auto portIndex = *magic_enum::enum_index ((Port) portType);
+                                         portTypes[portIndex] = mapper ((Port) portType); });
+    return portTypes;
+}
+} // namespace base_processor_detail
+
 class BaseProcessor : private JuceProcWrapper
 {
 public:
     using Ptr = std::unique_ptr<BaseProcessor>;
 
+    template <typename Port>
+    static constexpr auto defaultPortMapper (Port)
+    {
+        return PortType::audio;
+    }
+
+    template <typename InputPort,
+              typename OutputPort,
+              typename InputPortMapper = decltype (&defaultPortMapper<InputPort>),
+              typename OutputPortMapper = decltype (&defaultPortMapper<OutputPort>)>
     BaseProcessor (const String& name,
                    ParamLayout params,
+                   InputPort,
+                   OutputPort,
                    UndoManager* um = nullptr,
-                   int nInputs = 1,
-                   int nOutputs = 1);
+                   InputPortMapper inputPortMapper = &defaultPortMapper<InputPort>,
+                   OutputPortMapper outputPortMapper = &defaultPortMapper<OutputPort>) : JuceProcWrapper (name),
+                                                                                         vts (*this, um, Identifier ("Parameters"), std::move (params)),
+                                                                                         numInputs (magic_enum::enum_count<InputPort>()),
+                                                                                         numOutputs (magic_enum::enum_count<OutputPort>()),
+                                                                                         inputPortTypes (base_processor_detail::initialisePortTypes<InputPort> (inputPortMapper)),
+                                                                                         outputPortTypes (base_processor_detail::initialisePortTypes<OutputPort> (outputPortMapper))
+    {
+        std::cout << "Creating processor: " << name << std::endl;
+        std::cout << "With input ports: " << std::endl;
+        for (size_t i = 0; i < inputPortTypes.size(); ++i)
+            std::cout << "    " << magic_enum::enum_name ((InputPort) i) << " (" << magic_enum::enum_name (inputPortTypes[i]) << ")" << std::endl;
+        std::cout << "With output ports: " << std::endl;
+        for (size_t i = 0; i < outputPortTypes.size(); ++i)
+            std::cout << "    " << magic_enum::enum_name ((OutputPort) i) << " (" << magic_enum::enum_name (outputPortTypes[i]) << ")" << std::endl;
```

Comment:
This is great for confirming that everything works, but we should remember to remove it before merging :)

Reply:
thanks, will remove before merging